FC-Chess/FC-Chess.cpp

```diff
@@ -107,28 +105,271 @@ void FCC::Chess::checkMovesHorizontal()
+    if (checkIsPosInBounds(&i, &j)) 
+    {
+        if (!checkIsFriendlyOnField(&i, &j)) posPossible[i][j] = true;
+    }
+}
+
+void FCC::Chess::checkMovesKing()
+{
+    // creating variables for x and y
+    uint8_t i, j;
+
+    // checking possible moves
+    i = selX - 1, j = selY + 1;
+    for (; i <= selX + 1; i++)
+    {
+        if (checkIsPosInBounds(&i, &j))
+        {
+            if (!checkIsFriendlyOnField(&i, &j)) posPossible[i][j] = true;
+        }
     }
+
+    i = selX - 1, j = selY - 1;
+    for (; i <= selX + 1; i++)
+    {
+        if (checkIsPosInBounds(&i, &j))
+        {
+            if (!checkIsFriendlyOnField(&i, &j)) posPossible[i][j] = true;
+        }
+    }
+
+    i = selX - 1, j = selY;
+    if (checkIsPosInBounds(&i, &j))
+    {
+        if (!checkIsFriendlyOnField(&i, &j)) posPossible[i][j] = true;
+    }
+
+    i = selX + 1, j = selY;
+    if (checkIsPosInBounds(&i, &j))
+    {
+        if (!checkIsFriendlyOnField(&i, &j)) posPossible[i][j] = true;
+    }
+}
+
+void FCC::Chess::checkMovesPawn()
+{
+    // black pieces
+    if (posColor[selX][selY] == FCC::PiecesColors::BLACK)
+    {
+        // creating variables for x and y
+        uint8_t i, j;
+
+        // checking possible moves
+        i = selX - 1, j = selY - 1;
+        for (; i <= selX + 1;)
+        {
+            if (checkIsPosInBounds(&i, &j))
+            {
+                if (checkIsEnemyOnField(&i, &j)) posPossible[i][j] = true;
+            }
+            i += 2;
+        }
+
+        i = selX, j = selY - 1;
+        if (checkIsPosInBounds(&i, &j))
+        {
+            if (checkIsPositionClear(&i, &j)) posPossible[i][j] = true;
+        }
+    }
+
+    // white pieces
+    if (posColor[selX][selY] == FCC::PiecesColors::WHITE)
+    {
+        // creating variables for x and y
+        uint8_t i, j;
+
+        // checking possible moves
+        i = selX - 1, j = selY + 1;
+        for (; i <= selX + 1;)
+        {
+            if (checkIsPosInBounds(&i, &j))
+            {
+                if (checkIsEnemyOnField(&i, &j)) posPossible[i][j] = true;
+            }
+            i += 2;
+        }
+
+        i = selX, j = selY + 1;
+        if (checkIsPosInBounds(&i, &j))
+        {
+            if (checkIsPositionClear(&i, &j)) posPossible[i][j] = true;
+        }
+    }
+}
+
+bool FCC::Chess::checkIsPawnChangingToQueen()
+{
+    //black pieces
+    if (posColor[selX][selY] == FCC::PiecesColors::BLACK)
+    {
+        if (selY == 0) return true;
+        return false;
+    }
+
+    // white pieces
+    if (posColor[selX][selY] == FCC::PiecesColors::WHITE)
+    {
+        if (selY == 7) return true;
+        return false;
+    }
+
+    // no piece at all
+    return false;
+}
+
+void FCC::Chess::resetAllowedMovement()
+{
+    for (uint8_t i = 0; i <= 7; i++)
+    {
+        for (uint8_t j = 0; j <= 7; j++)
+        {
+            posPossible[i][j] = false;
+        }
+    }
+}
+
+bool FCC::Chess::checkIsPositionClear(uint8_t* x, uint8_t* y)
```

Comment:
The `uint8_t` datatype is actually a `char` with the sice to 8 bits / 1 byte. A pointer hold memory addresses, which are simple numbers. Pointers always use up 4 byte (32 bits).

Usually you want to use Copy by reference (COW) from C++'s move semantics. It basically doesn't copies the content or creates a pointer, but creates a new variable that has re-uses the data stored by the original variables that was passed as parameter. This saves up the pointer allocation and doesn't require you to pass variables using `&` (it's memory address).

Also since the used datatype uses less memory than the pointer, suing a pointer here doesn't make sense since it requires more data and more time to be allocated.

Since you want to pass data with changed values (like the `selX + 1`) you can't use COW or pointers anyway since the calculated value must be stored into a variable in order to be used. So let's just use a normal `uint8_t` here to make the code clean and simple.
Using pointers and COW is a good practice, but having to create variables over and over again manually or clearing one and so on makes code complicated, so you generally use COW when possible, if you have to pass values directly or using a calculation in the parameter field, you just create a normal variable as parameter.
When you are using C or need to push back data to the passed variable, you use a normal pointer, but in C++ you don't have the needs for that if you do it right.

Reply:
Thanks for the information. I'll apply it in the next commit.